tools/jison/parserDefinitions.js

```diff
@@ -0,0 +1,193 @@
+      `${LICENSE}${contents.replace(
+        'parse: function parse',
+        SQL_STATEMENTS_PARSER_JSDOC + 'parse: function parse'
+      )}\nexport default sqlStatementsParser;\n`
+  },
+  hplsqlStatementsParser: {
+    sources: [`${PARSER_FOLDER}/jison/hplsqlStatementsParser.jison`],
+    targetJison: 'hplsqlStatementsParser.jison',
+    outputFolder: PARSER_FOLDER,
+    parserName: 'hplsqlStatementsParser',
+    afterParse: async contents =>
+      `${LICENSE}${contents.replace(
+        'parse: function parse',
+        SQL_STATEMENTS_PARSER_JSDOC + 'parse: function parse'
+      )}\nexport default hplsqlStatementsParser;\n`
+  }
+};
+
+/**
+ * Searches through the SQL_FOLDER and if a jison/structure.json file exists it considers it a parser
+ */
+const findParserSources = async () => {
+  const folders = await listDir(SQL_FOLDER);
+  const structureFiles = [];
+  for (const folder of folders) {
+    const outputFolder = `${SQL_FOLDER}/${folder}`;
+    const jisonFolder = `${outputFolder}/jison`;
+    const structureFile = `${jisonFolder}/structure.json`;
+    if (fileExists(structureFile)) {
+      structureFiles.push({ dialect: folder, outputFolder, jisonFolder, structureFile });
+    }
+  }
+  return structureFiles;
+};
+
+/**
+ * Identifies all the SQL parsers based on subfolders in SQL_FOLDER and adds them to parserDefinitions
+ */
+export const identifySqlParsers = async () => {
+  const parserSources = await findParserSources();
+  const foundDefinitions = {};
+  for (const parserSource of parserSources) {
+    const structure = JSON.parse(await readFile(parserSource.structureFile));
+    if (structure.autocomplete) {
+      foundDefinitions[`${parserSource.dialect}AutocompleteParser`] = createParserDefinition(
+        structure.autocomplete,
+        parserSource,
+        true,
+        structure
+      );
+    }
+    if (structure.syntax) {
+      foundDefinitions[`${parserSource.dialect}SyntaxParser`] = createParserDefinition(
+        structure.syntax,
+        parserSource,
+        false,
+        structure
+      );
+    }
+  }
+  return { ...FIXED_PARSER_DEFINITIONS, ...foundDefinitions };
+};
+
+const createParserDefinition = (
+  sources,
+  { dialect, outputFolder, jisonFolder },
+  autocomplete,
+  { lexer, imports }
+) => {
+  const parserName = `${dialect}${autocomplete ? 'AutocompleteParser' : 'SyntaxParser'}`;
+
+  const absoluteSources = sources.map(source => `${jisonFolder}/${source}`);
+
+  for (const source of absoluteSources) {
+    if (!fileExists(source)) {
+      throw new Error(
+        `Could not find the file '${source}' as defined in structure.json for ${dialect}`
+      );
+    }
+  }
+
+  return {
+    sources: sources.map(source => `${jisonFolder}/${source}`),
+    lexer: `${jisonFolder}/${lexer}`,
+    targetJison: `${outputFolder}/${parserName}.jison`,
+    sqlParser: autocomplete ? 'AUTOCOMPLETE' : 'SYNTAX',
+    parserName,
+    outputFolder,
+    afterParse: async contents =>
```

Comment:
Can we add some comments to explain why and what we are replacing here? Or split this up into functions with descriptive names.

Reply:
Yes! Makes sense, I'll add some details.